### PR TITLE
Use "devtools" automation protocol as fallback

### PR DESCRIPTION
**Pull request by christian-bromann**
## Proposed changes

Given that Puppeteer now supports Chrome, Firefox and Edge (Chromium) without having to install a browser it makes sense to use it as fallback in case a driver doesn't exist. This patch adds:

- support for Edge (Chromium) using the devtools automation protocol
- removal of `puppeteer-firefox` dependency (now lives directly within `puppeteer`)
- driver endpoint check and fallback to `devtools` if endpoint couldn't be reached

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

This is work in progress. It needs to be tested on Linux and Windows.

### CODE REVIEW

File: packages/webdriverio/src/utils.js [hunk 1]
```diff
@@ -414,3 +415,33 @@ export const enhanceElementsArray = (elements, parent, selector, foundWith = '$$
  * @param {string} automationProtocol
  */
 export const isStub = (automationProtocol) => automationProtocol === './protocol-stub'
+
+export const getAutomationProtocol = async (config) => {
+    /**
+     * don't modify automation protocol if hostname is set as it is very likely
+     * that "webdriver" will be used
+     */
+    if (config.hostname || (config.user && config.key)) {
```
Comment (mgrybyk) [inline packages/webdriverio/src/utils.js]: should we add config.port as well?

Comment (mgrybyk) [inline packages/webdriverio/src/utils.js]: should we add config.port as well?

File: packages/webdriverio/src/utils.js [hunk 2]
```diff
@@ -414,3 +415,33 @@ export const enhanceElementsArray = (elements, parent, selector, foundWith = '$$
...
+
+    /**
+     * make a head request to check if a driver is available
+     */
+    const driverEndpointHeaders = await new Promise((resolve) => {
+        const req = http.request({
+            method: 'GET',
+            host: 'localhost',
+            port: config.port || 4444,
+            path: '/status'
```
Comment (mgrybyk) [inline packages/webdriverio/src/utils.js]: isn't config.path ignored in such case?

Comment (christian-bromann) [reply packages/webdriverio/src/utils.js]: I will remove `config.port` - if nothing is specified we need to assume that the driver is on `localhost:4444`.

Comment (mgrybyk) [reply packages/webdriverio/src/utils.js]: Yeah, but path /status is not necessarily valid. Can it be /wd/hub/path ?

Comment (christian-bromann) [reply packages/webdriverio/src/utils.js]: `/status` is defined by the WebDriver spec and therefor has to be there

**Comment by christian-bromann** [comment]
@webdriverio/project-committers 

This is finally ready for review, here is the Gist of things that change with this patch:

- I updated the Travis build to run lint, unit and smoke test separated from e2e (given that they require a different setup where browser need to be installed)

Build before:
Screenshot 2020-02-15 at 18 37 54

Build now:
Screenshot 2020-02-15 at 18 37 36

Given that we have 5 concurrent VMs the overall time shouldn't change much.

- I added support for Chromium Edge (currently only Mac and Windows)
- The DevTools automation is now default, that means if WebdriverIO can't find a driver endpoint it defaults to that
- due to a weird bug in Jest I had to change the setup for test coverage which now always includes all files (even the ones that are not test)

File: .travis.yml [hunk 3]
```diff
@@ -1,18 +1,91 @@
 dist: xenial
 language: node_js
 
-node_js:
-  - 10
-  - 12
+cache:
+  directories:
```
Comment (mgrybyk) [inline .travis.yml]: do we need to cache website's node modules here?

**Comment by mgrybyk** [comment]
is it due to Jest' bug there are some index files ignored `/* istanbul ignore file */` ?

File: packages/devtools/package.json [hunk 4]
```diff
@@ -35,12 +35,9 @@
     "@wdio/protocols": "6.0.0-alpha.1",
     "@wdio/utils": "6.0.0-alpha.1",
     "chrome-launcher": "^0.11.1",
-    "puppeteer-core": "^1.18.1"
+    "puppeteer-core": "^2.1.1"
   },
   "devDependencies": {
     "@types/puppeteer": "^1.19.0"
```
Comment (mgrybyk) [inline packages/devtools/package.json]: let's update types to 2.0.0 ?

File: packages/webdriverio/src/constants.js [hunk 5]
```diff
@@ -422,3 +421,9 @@ export const APPIUM_CAPABILITES = [
     ...APPIUM_ANDROID_CAPABILITIES,
     ...APPIUM_IOS_CAPABILITIES
 ]
+export const DRIVER_DEFAULT_ENDPOINT = {
+    method: 'GET',
+    host: 'localhost',
+    port: 4444,
+    path: '/status'
```
Comment (mgrybyk) [inline packages/webdriverio/src/constants.js]: path should respect the `path` defined by user in config! For selenium v3 and some cloud providers it should be /wd/hub/status

I've just verified it locally

File: e2e/testrunner.test.js [hunk 6]
```diff
@@ -1,10 +1,28 @@
...
+    const hasPassed = failures === 0
+
+    /**
+     * print log files for debugging if test fails
+     */
+    if (!hasPassed) {
+        const rootPath = path.join(__dirname, 'wdio')
+        const logFiles = fs.readdirSync(rootPath).filter((file) => (
+            // only log files
+            file.endsWith('.log')
```
Comment (erwinheitzman) [inline e2e/testrunner.test.js]: This filter does not work as `file.endsWith('.log')` needs to be returned here.

File: .travis.yml [hunk 7]
```diff
@@ -1,18 +1,93 @@
...
+    - node_modules
+    - packages/devtools/node_modules
```
[the remark this answers is not in the document]
Comment (christian-bromann) [reply .travis.yml]: well we don't want to cache our build files

File: e2e/testrunner.test.js (hunk 6 above)
Comment (christian-bromann) [reply e2e/testrunner.test.js]: it returns because the surrounding `(` and `)`

Comment (christian-bromann) [reply e2e/testrunner.test.js]: made it more clear

File: packages/webdriverio/src/constants.js (hunk 5 above)
Comment (christian-bromann) [reply packages/webdriverio/src/constants.js]: by default all drivers should start with path `/` if you change that we should assume you run `webdriver`. I modified the `getAutomationProtocol` method to now take this into consideration:

```js
/**
 * run WebDriver if hostname or port is set
 */
if (config.hostname || config.port || config.path || (config.user && config.key)) {
    return 'webdriver'
}
```

**Comment by christian-bromann** [comment]
@mgrybyk @erwinheitzman addressed your comments

**Comment by christian-bromann** [comment]
Moving forward as mentioned in the project committers channel.

**Pull request merged**